### PR TITLE
Introduces BulkVectorScorer for ExactSearch

### LINKED ISSUE #3348: [RFC] Simplify exact search to return a lucene Scorer
## Description
Currently, ExactSearcher - a construct that performs brute force k-NN search returns [TopDocs](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java#L94) which holds the top-k documents. This breaks the lazy flow of lucene's query->weight->scorer approach.

One of the drawbacks of returning TopDocs is that it needs to eagerly iterate through all the documents while maintaining the data structures to hold the topDocs to again wrap it in a scorer via either KNNScorer or DocAndScoreQueryScorer. 

For N segments in a shard there is
1. It create N+1 priority queues (Rescoring needs N+1 while KNNquery might not need it)
2. ScoreDoc[] arrays allocated per segment, then merged into another array
3. Callable + List<Callable> + task executor overhead per query

This creates a GC overhead of short lived objects. This is redundant because lucenes [TopScoreDocCollector](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L94) collects the results again with the same priority queue approach with an added benefit of early termination. 

## Proposal
The approach recommended is simple - Drop top-k calculations for exact search and return a scorer instead

### Consequences
**Rescoring**
Exact search is heavily used in rescoring, this will move all rescoring to lazy approach rather than eager score computations.
This means we loose aggressive parallelism and honor slice counts; both of which will have pros and cons

**Won't honor k value**
Current approach of k results won't be honored for any exact search. Theoretically this makes sense since all exact search is expected to loop though all results, having an additional loop to limit it to k value does not makes sense. Note that it will directly honor size values

**Refactoring**
This can involve significant refactoring

Sample implementation for RescoreKNNVectorQuery
- [BulkVectorScorer](https://github.com/shatejas/k-NN/commit/efe2a6e5aa3ad70a99f484c0178c34958562d058)
- [ExactSearcher](https://github.com/shatejas/k-NN/commit/7999c40fed18669b49354f01d8fd00c256755384)

Currently I don't have any benchmarks around this. Needed initial feed back while I benchmark in parallel

### PR DESCRIPTION
**Pull request by shatejas**
### Description

Introduces BulkVectorScorer, a Lucene Scorer implementation that wraps the bulk vector scoring API and encapsulates score filtering logic. This replaces the manual buffer-iteration loops previously spread across ExactSearcher methods, consolidating all exact-search scoring into a single reusable abstraction.

#### Key changes
- **New BulkVectorScorer class** — A Scorer backed by VectorScorer.Bulk that iterates documents in batches, applies a score predicate, and exposes standard 
- DocIdSetIterator semantics (nextDoc, advance, cost). Two factory methods express intent:
    - forKSearch — accepts all scores (used for top-k collection)
    - forRadialSearch — filters to scores ≥ a minimum threshold
- **Simplified ExactSearcher** — Removed searchTopK, searchWithMinScore, and the raw buffer loop in scoreAllDocs. All paths now construct the appropriate BulkVectorScorer and pass it to collectTopK or the simplified scoreAllDocs(Scorer). This eliminates duplicated iteration logic and makes the scoring strategy explicit at the call site.

### Test plan

  - [x] New BulkVectorScorerTests — covers k-search scoring, radial search filtering, filtered doc iteration, advance() semantics (including at/past target),
  cost, initial state, empty docs, and edge cases
  - [x] New ExactSearcherTests — unit tests for exact searcher flows
  - [x] Existing integration tests (MOSFaissFloatIndexIT, FilteredSearchANNSearchIT) validate end-to-end behavior

### Related Issues
Partially Resolves [#3348](https://github.com/opensearch-project/k-NN/issues/3348)

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Vikasht34 — COMMENTED** [review]
The refactoring is clean and the BulkVectorScorer abstraction makes the code much more readable Thanks @shatejas , I have one concern though , our Old code https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java#L227-L248 , we had optimization to skip block if score is less than max score 

```
if (maxScore < topDoc.score) {
    continue;  
}
```

Seems like now we are just using wrapper againt Lucene VectorScorer.Bulk interface that logic is missed . @navneet1v that skiping logic for meant for optimizating latency if I am nit wrong?

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java [hunk 1]
```diff
@@ -116,6 +119,34 @@ public TopDocs searchLeaf(final LeafReaderContext leafReaderContext, final Exact
...
+    public Scorer exactSearchScorer(final LeafReaderContext leafReaderContext, final ExactSearcherContext context) throws IOException {
+        final SegmentReader reader = Lucene.segmentReader(leafReaderContext.reader());
+        final FieldInfo fieldInfo = FieldInfoExtractor.getFieldInfo(reader, context.getField());
+        if (fieldInfo == null) {
+            log.debug("[KNN] FieldInfo is null for field [{}] in segment [{}]", context.getField(), reader.getSegmentName());
+            return null;
+        }
+
+        final VectorScorer vectorScorer = createVectorScorer(reader, fieldInfo, leafReaderContext, context);
+        assert vectorScorer != null;
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java]: any change vectorScorer will be null? if no then do we need this assert?

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java [hunk 2]
```diff
@@ -116,6 +119,34 @@ public TopDocs searchLeaf(final LeafReaderContext leafReaderContext, final Exact
...
+
+        // When nested, matchedDocsIterator is already consumed inside NestedBestChildVectorScorer,
+        // so pass null to avoid double consumption of the same iterator.
+        final DocIdSetIterator matchedDocs = context.getParentsFilter() != null ? null : context.getMatchedDocsIterator();
+
+        if (context.getRadius() != null) {
+            assert extractKNNEngine(fieldInfo) == KNNEngine.FAISS : "Exact searcher for Radial search is only used by FAISS engine";
+            final float minScore = context.isMemoryOptimizedSearchEnabled
+                ? context.getRadius()
+                : KNNEngine.FAISS.score(context.getRadius(), getSpaceType(modelDao, fieldInfo));
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:144]: is this change algined with what we have been doing in the radial search bug fix for IP space type? Ref: https://github.com/opensearch-project/k-NN/pull/3369

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java [hunk 3]
```diff
@@ -116,6 +119,34 @@ public TopDocs searchLeaf(final LeafReaderContext leafReaderContext, final Exact
         return exactNearestNeighborSearch(context, vectorScorer, matchedDocs);
     }
 
+    public Scorer exactSearchScorer(final LeafReaderContext leafReaderContext, final ExactSearcherContext context) throws IOException {
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:122]: I am not seeing use of this function, any reason for keeping it public like you want to use it in next PRs?

File: src/test/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorerTests.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+import org.apache.lucene.util.FixedBitSet;
+import org.opensearch.knn.KNNTestCase;
+import org.opensearch.knn.index.vectorvalues.TestVectorValues;
+
+import java.io.IOException;
+import java.util.List;
+
+public class BulkVectorScorerTests extends KNNTestCase {
+
+    private static final VectorSimilarityFunction SIMILARITY = VectorSimilarityFunction.EUCLIDEAN;
```
Comment (navneet1v) [inline src/test/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorerTests.java]: can we please add it for COSINE and inner product?

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 3 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:122]: correct, this is for integration with knn weight

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 1 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java]: searchLeaf has sort of a[ silent failure here ](https://github.com/opensearch-project/k-NN/pull/3361/changes#diff-506dbbeb9385167fbc3e3943521b79b918b294deda5764246804fbdd307a56e1R108) I can return an empty scorer to keep status quo

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 2 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:144]: Its the same as https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java#L180 which is not changed in #3369

**Comment by shatejas** [comment]
> we had optimization to skip block if score is less than max score

[This is where ](https://github.com/opensearch-project/k-NN/pull/3361/changes#diff-3a2a3a5e0a3cd3ff857b92e07602d1c7184a320aa42d62de222a63399d2db2acR73) its being handled, where it will reset the currentBatchIndes to the size, so it skips scanBufferForMatch() and bulk scores the next batch

File: src/test/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorerTests.java (hunk 4 above)
Comment (shatejas) [reply src/test/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorerTests.java]: Will do

**Comment by Vikasht34** [comment]
> > we had optimization to skip block if score is less than max score
> 
> [This is where ](https://github.com/opensearch-project/k-NN/pull/3361/changes#diff-3a2a3a5e0a3cd3ff857b92e07602d1c7184a320aa42d62de222a63399d2db2acR73) its being handled, where it will reset the currentBatchIndes to the size, so it skips scanBufferForMatch() and bulk scores the next batch

Got it ... Thanks

**Review by Vikasht34 — DISMISSED** [review]
LGTM ...

File: src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    }
+                    float maxBatchScore = bulkScorer.nextDocsAndScores(DocIdSetIterator.NO_MORE_DOCS, null, buffer);
+                    currentBatchIdx = 0;
+                    if (buffer.size == 0) {
+                        return currentDocId = NO_MORE_DOCS;
+                    }
+                    if (!scoreFilter.test(maxBatchScore)) {
+                        currentBatchIdx = buffer.size;
+                    }
+                }
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java:77]: @shatejas can we add the logic of skipping the whole set of vectors if the max score < minScore we have seen till now during the iteration?

**Comment by navneet1v** [comment]
> The refactoring is clean and the BulkVectorScorer abstraction makes the code much more readable Thanks @shatejas , I have one concern though , our Old code https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java#L227-L248 , we had optimization to skip block if score is less than max score
> 
> ```
> if (maxScore < topDoc.score) {
>     continue;  
> }
> ```
> 
> Seems like now we are just using wrapper againt Lucene VectorScorer.Bulk interface that logic is missed . @navneet1v that skiping logic for meant for optimizating latency if I am nit wrong?

Some clarification all that logic is doing was it will skip the block of vectors to be passed through topK queue. @Vikasht34 good catch. I added a comment here: https://github.com/opensearch-project/k-NN/pull/3361/changes#r3483691142

File: src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java (hunk 5 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java:77]: https://github.com/opensearch-project/k-NN/pull/3361#issuecomment-4812413522

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java:77]: @shatejas I don't see that `scoreFilter` getting updated when maxscore is change. I might be missing the code where you are updating the score. can you please point me to that?

Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/BulkVectorScorer.java:77]: I misunderstood when @Vikasht34 pointed out, I thought it was for radial. Added support for minCompetitive score

**Review by Vikasht34 — DISMISSED** [review]

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 3 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:122]: I didn't get it. If this is for integration with KNNWeight I don't see that integration. Little confused here.

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java [hunk 6]
```diff
@@ -172,121 +203,56 @@ private TopDocs doRadialSearch(
...
+        for (int doc = iter.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = iter.nextDoc()) {
+            float score = scorer.score();
+            if (score > topDoc.score) {
+                topDoc.score = score;
+                topDoc.doc = doc;
+                topDoc = queue.updateTop();
+                collectedCount++;
+                if (updateMinCompetitiveScore && collectedCount >= heapSize) {
+                    scorer.setMinCompetitiveScore(topDoc.score);
                 }
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:259]: why not bulkVectorScorer encapsulating this score change. Won't that be better if bulkVectorScorer has this and ensure that only return docs which are better than the minCompetitive score it has seen.

Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:259]: It won't work across segments, The collector is kept responsible for minCompetitive score to be able to have it updated for scorers across segments and not keep it localized, within exact searcher its being honored locally but if a top level collector wants to update the minCompetitive score it won't be able to 

The Scorable interface has setMinCompetitiveScore to be able to leverage it across segments and not just locally; its designed keeping multi-segment use case in mind

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 3 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:122]: Yeah it will be in consecutive PRs

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 6 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:259]: @shatejas I am not sure if I got the explanation here. 

Let me check with you if my understanding is correct:

BulkScorer is at a segment level right? It should track its own minCompetitive Score which it is doing right now. So is the idea being Scorer will keep the track, but it will never take a decision whether to collect a doc or not. The collection logic based on minCompetitive score will be handled by the collector, in this case it ExactSearcher class.

If its correct then, I am good here.

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 1 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java]: I think its better log the error if its truely an error and return an empty scorer rather than failing on asserts.

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 6 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java:259]: > So is the idea being Scorer will keep the track, but it will never take a decision whether to collect a doc or not. The collection logic based on minCompetitive score will be handled by the collector, in this case it ExactSearcher class.

Correct. The idea here is apart from ExactSearcher class, [TopScoreDocCollector](https://github.com/apache/lucene/blob/8fe701c0873de37b95fabd4a0feafa2e16f30097/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L162) or any other collector, should be able to leverage it, the minCompetitve score for TopScoreDocCollector is updated locally as well as [gloabaly](https://github.com/apache/lucene/blob/8fe701c0873de37b95fabd4a0feafa2e16f30097/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L89C11-L89C42) that is across segments

File: src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java (hunk 1 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/exactsearch/ExactSearcher.java]: I changed it to throw IllegalState to avoid silent failures, there is no case where vector scorer is null

**Comment by navneet1v** [comment]
@shatejas can you please use `./gradlew spotlessApply`

**Review by navneet1v — APPROVED** [review]

**Review by Vikasht34 — APPROVED** [review]
